packages/rollup-contracts/contracts/state-tree/FullPatriciaTree.sol

```diff
@@ -0,0 +1,380 @@
+        D.Label memory remaining;
+        // We initialize to the full key
+        remaining = D.Label(key, 256);
+        // Keeps track of how much we have "eaten" into the key.
+        // It should always hold that bitsTraversed + remaining.length == keyLength (256) at the end of each loop iteration
+        uint bitsTraversed = 0;
+        // Current edge the traversal is processing.
+        // Each loop iteration will chose the right or left child as the new current edge.
+        D.Edge memory currentEdge;
+        // Start traversal at the root
+        currentEdge = tree.rootEdge;
+
+        // Proof to return along with branch bitmask
+        bytes32[256] memory siblings;
+        uint numSiblings = 0;
+        while (true) {
+            // Figure out the common prefix between the current edge and remaning bits in the traversal.
+            // If the requested key has indeed been set, the current edge should be a prefix of the remaining.
+            D.Label memory prefix;
+            D.Label memory suffix;
+            (prefix, suffix) = Utils.splitCommonPrefix(remaining, currentEdge.label);
+            require(
+                prefix.length == currentEdge.label.length,
+                'Reached an edge in traversal whose label is not a strict prefix of the remaining part of key.  This indicates that the requested key has not been set.'
+            );
+            if (suffix.length == 0) {
+                // Found a match!
+                break;
+            }
+            // Now that we are traversing this edge, add its length to bitsTraversed.
+            bitsTraversed += prefix.length;
+            // The next bit in the key determines whether to branch left or right.
+            // So, this sets the bitsTraversed'th bit in the branch mask to 1.
+            branchMask |= uint(1) << (255 - bitsTraversed);
+
+            // As explained in the last line, we traverse left or right based on the next bit.
+            uint head;
+            D.Label memory tail;
+            (head, tail) = Utils.chopFirstBit(suffix);
+            // head, either 0 or 1, tells us which child edge is to traverse to, and which is the sibling to supply in our proof.
+            uint siblingIndex = 1 - head;
+            siblings[numSiblings++] = edgeHash(
+                tree.nodes[currentEdge.node].children[siblingIndex]
+            );
+            // Now, update the current edge to be processed in next iteration
+            currentEdge = tree.nodes[currentEdge.node].children[head];
+            // Account for having processed another bit by choosing left or right.
+            bitsTraversed += 1;
+            remaining = tail;
+        }
+        if (numSiblings > 0)
+        {
+            _siblings = new bytes32[](numSiblings);
+            for (uint i = 0; i < numSiblings; i++)
+                _siblings[i] = siblings[i];
+        }
+    }
+
+    function getNonInclusionProof(Tree storage tree, bytes32 key) internal view returns (
```

Comment:
Could you add a function header? There's a lot going on in the return types, so it could be useful to detail what they represent.